### ISSUE TITLE
setHeight + redraw Seem to Break Virtual Dom Rendering

### ISSUE THREAD

**Issue by xXmrotsXx** [olifolkerd/tabulator#869]
Hi again @olifolkerd,

_TL;DR: I have actually solved (worked around?) my issue, but would appreciate your feedback on what I have documented below. Thanks again for all the support you provide._

Being new to tabulator I'm still getting my head around the inner workings of the code. I'm loving what I've been able to learn in just a couple of days using the product so thanks for it and the awesome support you provide! 

I'm now running into issues when dynamically setting the table height which I realise needs to be set for the Virtual DOM to function. 

1. The docs state _"...it can be set to any valid height css value."_ EG:
```
$("#example-table").tabulator({
    height:"100%", //set the table height option
});
```
However, this appears to render the table height to 100% of the DATA, not the available screen height like it would in CSS. Is this the correct functionality or is something going wrong? Also, only the first 20 rows have been populated...

2. To get around the problems I was having with 1 above, I used the following:
```
$("#example-table").tabulator("setHeight", $(window).height() - 250);
$("#example-table").tabulator("redraw", true); 
```
I tried using it in place of the height assignment above, but cannot use _setHeight_ until the table has been constructed. So I moved it outside of the table constructor and thought my issues had been resolved... BUT, the Virtual DOM still did not update and my table only populated 50 rows this time.

3. Leaving 2 in place I can set height: to anything >0 (1px, 800px, 100%) and all my problems go away... Why? :-)

Thanks again

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

the ***height*** property just sets the height CSS value of the table to the given value, and then uses the value to calculate the virtual dom height so it should fill the container. that suggests that the containing element has no height set on it.

You shouldnt need to call ***redraw*** after setting the height of the Table Tabulator should do that for you, it is likely that JS is trying to call the redraw function before the the table has finished rendering, which may be confusing things.

Have you tried building your table on a clean HTML page with no other code just to check that  nothing else on the page is interfering with it? If you are still having trouble after that, post a copy of your table constructor and i will see what i can suggest.

Cheers

Oli :)

**Comment by xXmrotsXx** [comment]
Hey Oli,

Thanks for the feedback. I won't have a chance to look at it until tomorrow, but the reason I call redraw was because the table population was agonisingly slow on sorts without it.

So are you saying it is right to have to have a CSS height for the table to be populated, then I use the setheight after to responsively resize it?

Thanks again

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

I would suggest that you wait till 3.4 is release tomorrow it has a more advance virtual DOM and doesn't need the height set through the table anymore but it will need to be set in CSS.

If a table dosnt have a height set, then t will render using standard mode which is very slow, the table must have a height to enable the virtual DOM.

What you are describing suggest to me that the other elements on your page that contain the table dont have their heights set correctly, so the 100% isnt doing what you would expect.

Can you provide a copy of the source for your page, on a CodePen example and i can give you some more feedback.

Cheers

Oli :)

**Comment by xXmrotsXx** [comment]
Hey @olifolkerd,

I've had a quick play around with 3.4, but my issues above have not been resolved... Keeping in mind that I had a work around in 3.3 anyway, 3.4 actually makes my table worse - most likely because of the dodgy way I've set it up ;-)

 1. Side by side comparison with my code as is - 3.4 actually renders the initial table about 10-20% slower.

2. Remove redraw as suggested - unusable - 10-20 seconds to load.

I've got a codepen link, but I don't have a paid account to attach my actual CSV file that is populating the table.

CP: https://codepen.io/xXmrotsXx/pen/vdmaVg
Search Icon: https://drive.google.com/open?id=12f1z2lTdpptp6sTfY3ZbB7beIWsxhiNW
Data: https://drive.google.com/open?id=1tp6QStsN4gw7_jDNmoOA7TP07s5QiEg-

Sorry in advance for what is most likely a rookie error.

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

Thanks for sending that over, there were some issues with your codepen setup, but i coped the code to my computer and got it up and running, the issues are with table height.

I would Strongly suggest moving to 3.4 as it fixes a number of issues in 3.3.

Some pointers on the code (please trust me and make the following changes):
- you need to remove the height property from your tabulator constructor, and the ***setHeight*** function and the ***redraw*** after it.
- you need to add the following line to your CSS which will set the height correctly `.tabulator{	height:calc(100% - 250px);}`
- you should never make global adjustements like the ` * {box-sizing:border-box}` as they will interfere with the styling of external libraries which rely on expected browser defaults.

With those changes I had the table displaying properly and loading fast.

Let me know how you get on.

Cheers

Oli

**Comment by olifolkerd** [comment]
I should have mentioned the tips above are for version 3.4.

If you want to keep using 3.3 you should follow the same steps, but dont set the height in the CSS, instead set it in the tabulator constructor to this:
```js
height:"calc(100% - 250px)",
```

Cheers

Oli :)

**Comment by xXmrotsXx** [comment]
Thanks so much again. I'll implement these changes and report back.

**Comment by xXmrotsXx** [comment]
Hi @olifolkerd,

I finally had another look at your suggestions, but unfortunately it didn't help - Using 3.4 my load/filters/sorts are still very slow. I've attached grabs of the Chrome Performance tab for you to see...

3.3 using my hacks (1.5s page load): [pic](https://drive.google.com/open?id=1tvKE4OUD5bxwNscrON_iW0xDngvVekUtl)
3.4 using your changes (7.4s page load): [pic](https://drive.google.com/open?id=1ZhgxKG3Y0u-uGyr1pp2pRs_SskXz8N32)

On a side note, replacing my height in the table constructor and removing the **_setheight_** and **_redraw_** (in 3.3) still only renders 20 rows and makes the actual table the height of the data in the text file.

For all the reasons you've mentioned, I would like to get it working with 3.4, but I can't explain how yours was working well and my 3.4 implementation so slow...

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

Sorry to hear you are still having issues,

looking at your screenshot for my proposed solution it is clear that the height is not correctly set on the table as the scroll bar for the page is enormous, it must therefore be in standard rendering mode, which is very slow for large data sets

I can guarantee you 100% that you do not need to use the setHeight function unless you need to force a resize of the table, which you shouldn't if the CSS is correctly configured, you just have to set a height on the table, and as i have said, i think there are issues with the rest of your page setup that are causing the issue.

To get to the bottom of this I would suggest that you create a table on an empty HTML page, with the body of the page containing the the div for the table and nothing else. and only include the libraries and CSS needed to make tabulator function, then set a height on that div in CSS and create your table without ***setHeight*** and see if it works.

If it does then you should add the elements of your page back in one at a time until you isolate where the problem is coming from

I hope that helps,

Cheers

Oli :)

**Comment by xXmrotsXx** [comment]
I feel so bad to keep bothering you for what I feel should be such a trivial matter... I think I have a fundamental flaw in my height setting. 

I don't know what is going on, but I started from scratch as you suggested, and even manually set the tabledata array with ~200 rows. No other scripts or CSV conversions to get the data in... As you can see [here](https://drive.google.com/open?id=1y7rM8qySlp9IO7nr-1C17U0xfw8PzQt9), my height is not being set and the table continues off screen.

I've added `.tabulator{height:205;}` to the bottom of `tabulator_site.css` and not used my personal overrides, but adjusting it has no effect. (height is still in the table constructor of the website documentation btw.)

I'm sure I've included everything I need in the <head>, and done the **_require()_** although commenting it out seems to have no effect. On a side note, why do we include stuff, but then have to _require_ it too?

My FreshStart html is below. I'm hoping something pops out at you...
```
<!DOCTYPE html>
<html>
	
	<head>
		<script type="text/javascript" src="node_modules/jquery/dist/jquery.min.js"></script> <!-- from the guide -->
		<script type="text/javascript" src="node_modules/jquery-ui-stable/jquery-ui.min.js"></script> <!-- from the guide -->
		<link href="node_modules/jquery.tabulator/dist/css/tabulator_site.css" rel="stylesheet"> <!-- Default Tabulator Stylesheet -->
		<script type="text/javascript" src="node_modules/jquery.tabulator/dist/js/tabulator.min.js"></script> <!-- from the guide -->
		<script type="text/javascript" src="node_modules/jquery-ui/external/requirejs/require.js"> <!-- I added this as the next function depends on it -->
			require("node_modules/jquery.tabulator"); 
		</script>
	</head>
	<body>
		<div id="example-table"></div>
		<script>
			//create Tabulator on DOM element with id "example-table"
			$("#example-table").tabulator({
				//height:205, // set height of table (in CSS or here), this enables the Virtual DOM and improves render speed dramatically (can be any valid css height value)
				layout:"fitColumns", //fit columns to width of table (optional)
				columns:[ //Define Table Columns
					{title:"Player", field:"Player", frozen:true},
					{title:"Age", field:"Age", align:"center"},
					{title:"Club", field:"Club", align:"center"},
					{title:"Pos", field:"Pos"},
					{title:"17 Avg", field:"17 Avg", align:"right", cssClass:"avgs", sorter:"number"},
					{title:"17 GP", field:"17 GP", align:"right"},
					{title:"17 Mins", field:"17Min", align:"right"},
					{title:"17 PPM", field:"17PPM", align:"right"},
					{title:"16 Avg", field:"16 Avg", align:"right", cssClass:"avgs", sorter:"number"},
					{title:"16 GP", field:"16 GP", align:"right"},
					{title:"16 Mins", field:"16Min", align:"right"},
					{title:"16 PPM", field:"16PPM", align:"right"},
					{title:"15 Avg", field:"15 Avg", align:"right", cssClass:"avgs", sorter:"number"},
					{title:"15 GP", field:"15 GP", align:"right"},
					{title:"15 Mins", field:"15Min", align:"right"},
					{title:"15 PPM", field:"15PPM", align:"right"},
					{title:"14 Avg", field:"14 Avg", align:"right", cssClass:"avgs", sorter:"number"},
					{title:"14 GP", field:"14 GP", align:"right"},
					{title:"14 Mins", field:"14Min", align:"right"},
					{title:"14 PPM", field:"14PPM", align:"right"},
				],
				rowClick:function(e, row){ //trigger an alert message when the row is clicked
					alert("Row " + row.getData().id + " Clicked!!!!");
				},
			});
			//define some sample data
			var tabledata = [
				{Player:"Cam Smith", Age:"36", Club:"MEL", Pos:"HOK", "17 Avg":"60"},
				{Player:"Cam Smith", Age:"36", Club:"MEL", Pos:"HOK", "17 Avg":"60"},
				... x ~200...
				{Player:"Cam Smith", Age:"36", Club:"MEL", Pos:"HOK", "17 Avg":"60"},
			];
			//load sample data into the table
			$("#example-table").tabulator("setData", tabledata);
			</script>
	</body>
</html>
```

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

Its no problem, we will get to the bottom of the issue.

I havnt had a chance to look through all your code, but as a start for you, if you are using node and you have installed Tabulator in your project then you can just use the ***require*** statement, you dont need the script tags to, and the same goes for jquery and jquery ui

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

Try downloading this mornings patch release 3.4.2 and see if that fixes things.

Cheers

Oli :)

**Comment by xXmrotsXx** [comment]
Hi @olifolkerd,

Unfortunately there is no noticeable difference. The Height, Virtual DOM nor the offset from the CSS seems to be getting applied...

**Comment by olifolkerd** [comment]
Is your table hidden when it is built?

**Comment by xXmrotsXx** [comment]
No, you can see the test table here: 

https://drive.google.com/open?id=1y7rM8qySlp9IO7nr-1C17U0xfw8PzQt9

**Comment by olifolkerd** [comment]
im confused, that appears to have rendered correctly

**Comment by xXmrotsXx** [comment]
The bottom of the table is off the screen, and changing the height in CSS has no effect. I was assuming it's still rendering in standard mode...

**Comment by olifolkerd** [comment]
it looks like it is rendering in standard mode.

Is there any chance you could work out the minimum configuration of table that this happens in for you (no additional formatters or any other code that could complicate the example etc) and send over the full HTML file, it might give me a fighting chance of seeing where you issue is coming from.

Cheers

**Comment by xXmrotsXx** [comment]
This table is just that... The html is above. I've just removed the middle 200 rows of manually entered data in my pasted version. I used the tabulator site CSS with the height in .tabulator at the bottom...

**Comment by olifolkerd** [comment]
you appear to be pulling in two versions of the jquery library, which you shouldnt need to

**Comment by olifolkerd** [comment]
for the purposes of testing, would you be able to do it by referencing all your jquery and tabulator libraries from the listed CDN's that way i can be 100% sure that we are operating on the same code base, i will then take a screenshot of how it renders on my computer and if you could then post the same we can compare

**Comment by xXmrotsXx** [comment]
Oh ok. I don't think I could find require in the library that I first downloaded... I thought that was normal. I'll have a look at that.

Thanks again, you are spending alot of time with this and I daresay it will end up being something stupid I've done. ;)

**Comment by olifolkerd** [comment]
no problem, it may well be an issue with the new rendering engine, and if it isnt it will still be good for others to know how the problem got solved.

**Comment by olifolkerd** [comment]
have a look at the answer to the issue i just posted on #887 i think it might be relevant to you

**Comment by xXmrotsXx** [comment]
OK, so my head now looks like this:
```
<head>
	<meta name="viewport" content="width=device-width, initial-scale=1">	
	<script type="text/javascript" src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
	<script type="text/javascript" src="https://code.jquery.com/ui/1.12.1/jquery-ui.min.js"></script>
	<!-- STYLESHEETS... CHANGE THE STYLE HERE -->
	<link href="node_modules/jquery.tabulator/dist/css/tabulator_site.css" rel="stylesheet"> <!-- Default Tabulator Stylesheet -->
	<link href="css/mainstyle-min.css" rel="stylesheet">                                     <!-- My Override Stylesheet -->
		
	<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/tabulator/3.4.2/js/tabulator.min.js"></script>
		
</head>
```
I believe now after reading the other issue as well, that my problem is setting the height. I stopped using a % and for this test just used a fixed 500px.

Now this is where my issue could become a deeeerr moment, but I just can't see why this is happening... 

When I set the height in my table only example, using your tabulator_site.css stylesheet, the table seems to resize and the Virtual DOM is active. But if remove it from there - keeping yours default, and copy the same statement to my stylesheet (that I link after yours) there still seems to be no height set or Virtual DOM.

**Comment by xXmrotsXx** [comment]
UPDATE: Clearing the cache in Chrome seems to have fixed all my height problems.

- After a cache clear, setting the size of the table in px worked without needing to set html and body to 100% in CSS.

- html and body need to be set to 100% to use `height: calc(100% - 250px)` in CSS.

I'm not sure if it is something I did wrong or if tabulator or Chrome is caching some things they shouldn't be. I have run into the same problem a few times and clearing the cache always resolves the issue. I can't seem to reproduce it myself though unfortunately.

In summery, I'm happy to call this resolved! Yay! I'm currently using Tabulator 3.4.2, have removed my **_setHeight_**, **_redraw_** and **_require_** functions completely, and am only using one jQuery-ui library :-). After clearing the cache on Chrome and refreshing my page, all tables are now working perfectly.

Thanks so much for all of your time on this issue @olifolkerd, I really appreciate it. The quick start example still mentions setting the height in the table constructor, but I notice you have updated it a lot since I last looked. So is that still OK to do, or should it be in CSS?

**Comment by olifolkerd** [comment]
Hey @xXmrotsXx 

Great to hear that is all working for you now, you can set the height wherever you like as long as it is set somewhere, i only include it in the quick start to show it is a possibility, the intro to the virtual DOM section that has just been added to the quickstart guide goes into more detail about how to set the table height.

Cheers

Oli :)
